subprojects/logging/src/main/java/org/gradle/util/CollectionUtils.java

```diff
@@ -58,23 +59,41 @@
  * Plugins should prefer external collection frameworks over this class.
  * Internally, all code should use {@link org.gradle.util.internal.CollectionUtils}.
  *
- * @deprecated Will be removed in Gradle 8.0.
+ * @deprecated Will be removed in Gradle 9.0.
  */
 @Deprecated
 public abstract class CollectionUtils {
 
+    private static void logDeprecation() {
+        DeprecationLogger.deprecateType(CollectionUtils.class)
+            .willBeRemovedInGradle9()
+            .withUpgradeGuideSection(7, "org_gradle_util_reports_deprecations")
+            .nagUser();
+    }
+
+    public CollectionUtils() {
+        logDeprecation();
+    }
+
     /**
      * Returns null if the collection is empty otherwise expects a {@link #single(Iterable)} element to be found.
      */
     @Nullable
     public static <T> T findSingle(Iterable<T> source) {
-        return Iterables.isEmpty(source) ? null : single(source);
+        logDeprecation();
+        return Iterables.isEmpty(source) ? null : singleInternal(source);
     }
 
     /**
      * Returns the single element in the collection or throws.
      */
     public static <T> T single(Iterable<? extends T> source) {
+        logDeprecation();
+        return singleInternal(source);
+    }
+
+    private static <T> T singleInternal(Iterable<? extends T> source) {
+        logDeprecation();
```

Comment:
All calls are protected with logging. So this looks like a duplicate.
Shouldn't we either log here and not in the callers or not log here?

The later would match what you did for `collectInternal`.

Reply:
Nice catch. Did the later